bin/i18n/utils/crowdin_client.rb

```diff
@@ -0,0 +1,221 @@
+
+        if crowdin_file
+          request(
+            :update_or_restore_file,
+            crowdin_file['id'],
+            storageId: crowdin_storage_id,
+          )['data']
+        else
+          crowdin_directory = mutex.synchronize {source_directory File.dirname(file_path)}
+
+          request(
+            :add_file,
+            storageId: crowdin_storage_id,
+            directoryId: crowdin_directory&.dig('id'),
+            name: crowdin_storage['fileName'],
+          )['data']
+        end
+      ensure
+        request(:delete_storage, crowdin_storage_id) if crowdin_storage_id
+      end
+
+      private
+
+      attr_reader :project, :client
+
+      def mutex
+        @mutex ||= Mutex.new
+      end
+
+      def crowdin_source_path(source_path)
+        path = source_path.delete_prefix(CDO.dir).split(I18N_SOURCE_DIR).last || ''
+
+        path = File.join(File::SEPARATOR, path)
+        path = path.delete_suffix(File::SEPARATOR)
+
+        path
+      end
+
+      def crowdin_source_name(source_path)
+        File.basename(source_path).remove(File::SEPARATOR)
+      end
+
+      def source_directory(dir_path)
+        @source_directories ||= {}
+
+        crowdin_dir_path = crowdin_source_path(dir_path)
+        return @source_directories[crowdin_dir_path] if @source_directories.key?(crowdin_dir_path)
+
+        @source_directories[crowdin_dir_path] = get_source_directory(dir_path) || add_source_directory(dir_path)
+      end
+
+      def stringify_errors(errors)
+        messages = []
+
+        errors.each do |error|
+          error_key = error.dig('error', 'key')
+
+          error.dig('error', 'errors').each do |child_error|
+            error_code = child_error['code']
+            error_message = child_error['message']
+
+            messages << "#{error_key}[#{error_code}]: #{error_message}"
+          end
+        end
+
+        messages.join("\n")
+      end
+
+      def request(endpoint, *params)
+        response = client.public_send(endpoint, *params)
+
+        if response.is_a?(String) && response.include?('Something went wrong')
+          # e.g. "Something went wrong while request processing. Details - 429 Too Many Requests"
+          raise RequestError, response
+        elsif response.is_a?(Hash) && response['error']
+          # e.g. {"error"=>{"code"=>429, "message"=>"Too Many Requests"}}
+          raise RequestError, "#{response.dig('error', 'code')} #{response.dig('error', 'message')}"
+        elsif response.is_a?(Hash) && response['errors']
+          # e.g. {"errors"=>[{"error"=>{"key"=>"id", "errors"=>[{"code"=>"notNull", "message"=>"Can't be null"}]}}]}
+          raise RequestError, stringify_errors(response['errors'])
+        end
+
+        response
+      rescue StandardError => exception
+        attempt = (attempt || 0) + 1
+
+        if attempt <= REQUEST_RETRY_ATTEMPTS && RETRIABLE_ERRORS.any? {|error| exception.message.include?(error)}
+          sleep(REQUEST_RETRY_DELAY) && retry
```

Comment:
Why use `&&` here rather than
```suggestion
          sleep(REQUEST_RETRY_DELAY)
          retry
```
?